src/Models/Refund.php

```diff
@@ -0,0 +1,237 @@
+            if ($refund->amount > ($refund->payment->amount_refundable)) {
+                throw new \Exception('Please check the payment for the max amount that can be refunded.');
+            }
+            if (auth()->check()) {
+                $refund->updater_id = auth()->id();
+            }
+        });
+
+        static::created(function ($refund) {
+            $refund
+                ->payment
+                ->generateAmountRefunded()
+                ->save();
+        });
+    }
+
+    /**
+     * Issue refund.
+     *
+     * @return self
+     * @throws \Exception
+     */
+    public function issue()
+    {
+        switch ($this->method) {
+            case Refund::METHOD_STRIPE:
+                return $this->stripeRefund();
+
+                break;
+            case Refund::METHOD_VOUCHER:
+                return $this->voucherRefund();
+
+                break;
+        }
+    }
+
+    /**
+     * Is a stripe refund.
+     *
+     * @return bool
+     */
+    public function isStripe()
+    {
+        return $this->method == Refund::METHOD_STRIPE;
+    }
+
+    /**
+     * Is a voucher refund.
+     *
+     * @return bool
+     */
+    public function isVoucher()
+    {
+        return $this->method == Refund::METHOD_VOUCHER;
+    }
+
+    /**
+     * Notify customner about refund.
+     *
+     * @return void
+     */
+    public function notifyCustomer()
+    {
+        $this->payment->customer->user->notify(new RefundConfirmation($this));
+    }
+
+    /**
+     * Generate formated amount.
+     *
+     * @return string
+     */
+    public function decoratedAmount()
+    {
+        return '$' . number_format($this->amount / 100, 2, '.', ',');
+    }
+
+    /**
+     * Refund transaction to voucher.
+     *
+     * @return Refund
+     */
+    public function voucherRefund()
+    {
+        $amount = $this->amount;
+
+        $voucher = config('refunds.voucher_model')::create([
+            'location_id' => $this->payment->order->location_id,
+            'customer_id' => $this->payment->customer_id,
+            'voucher_type_id' => Refund::REFUND_VOUCHER_TYPE_ID,
+            'redeemable_at' => now(),
+            'amount' => $amount,
+            'creator_id' => $this->creator_id,
+            'updater_id' => $this->updater_id,
+        ]);
+
+        $this->fill([
+            'issued_at' => now(),
+            'issuer_id' => auth()->id(),
+            'voucher_id' => $voucher->id,
+        ]);
+
+        $this->save();
+
+        return $this;
+    }
+
+    /**
+     * Refund stripe payment.
+     *
+     * @return Refund
+     */
+    public function stripeRefund()
+    {
+        $options = [];
+
+        $payment = $this->payment;
+        $amount = $this->amount;
+
+        Stripe::setApiKey($payment->order->location->stripe_secret);
+
+        config(['cashier.key' => $payment->order->location->stripe_publishable]);
+        config(['cashier.secret' => $payment->order->location->stripe_secret]);
```

Comment:
This relationship chain - payment to order to location - is potentially straddling package boundaries.   It implies deep knowledge of payments and orders which are (currently) external.  I'm not positive the 3 (Orders, Payments and Refunds) can be split into different packages - they are all pretty tightly coupled to one another.

Reply:
I'm on the same page. I think the granular structure will generate a lot of complexity. I would make one large "Checkout" package instead but this is something that would need a discussion.

Reply:
Let's discuss in a Zoom meeting later today.